flink-runtime/src/main/java/org/apache/flink/runtime/state/heap/HeapSavepointStrategy.java

```diff
@@ -0,0 +1,115 @@
+
+/** A strategy how to perform a snapshot of a {@link HeapKeyedStateBackend}. */
+class HeapSavepointStrategy<K>
+        implements SnapshotStrategy<KeyedStateHandle, FullSnapshotResources<K>> {
+
+    private final Map<String, StateTable<K, ?, ?>> registeredKVStates;
+    private final Map<String, HeapPriorityQueueSnapshotRestoreWrapper<?>> registeredPQStates;
+    private final StreamCompressionDecorator keyGroupCompressionDecorator;
+    private final LocalRecoveryConfig localRecoveryConfig;
+    private final KeyGroupRange keyGroupRange;
+    private final StateSerializerProvider<K> keySerializerProvider;
+    private final int totalKeyGroups;
+
+    HeapSavepointStrategy(
+            Map<String, StateTable<K, ?, ?>> registeredKVStates,
+            Map<String, HeapPriorityQueueSnapshotRestoreWrapper<?>> registeredPQStates,
+            StreamCompressionDecorator keyGroupCompressionDecorator,
+            LocalRecoveryConfig localRecoveryConfig,
+            KeyGroupRange keyGroupRange,
+            StateSerializerProvider<K> keySerializerProvider,
+            int totalKeyGroups) {
+        this.registeredKVStates = registeredKVStates;
+        this.registeredPQStates = registeredPQStates;
+        this.keyGroupCompressionDecorator = keyGroupCompressionDecorator;
+        this.localRecoveryConfig = localRecoveryConfig;
+        this.keyGroupRange = keyGroupRange;
+        this.keySerializerProvider = keySerializerProvider;
+        this.totalKeyGroups = totalKeyGroups;
+    }
+
+    @Override
+    public FullSnapshotResources<K> syncPrepareResources(long checkpointId) {
+        return HeapSnapshotResources.create(
+                registeredKVStates,
+                registeredPQStates,
+                keyGroupCompressionDecorator,
+                keyGroupRange,
+                getKeySerializer(),
+                totalKeyGroups);
+    }
+
+    @Override
+    public SnapshotResultSupplier<KeyedStateHandle> asyncSnapshot(
+            FullSnapshotResources<K> syncPartResource,
+            long checkpointId,
+            long timestamp,
+            @Nonnull CheckpointStreamFactory streamFactory,
+            @Nonnull CheckpointOptions checkpointOptions) {
+        return new FullSnapshotAsyncWriter<>(
+                checkpointOptions.getCheckpointType(),
+                createCheckpointStreamSupplier(checkpointId, streamFactory, checkpointOptions),
+                syncPartResource);
+    }
+
+    private SupplierWithException<CheckpointStreamWithResultProvider, Exception>
+            createCheckpointStreamSupplier(
+                    long checkpointId,
+                    CheckpointStreamFactory primaryStreamFactory,
+                    CheckpointOptions checkpointOptions) {
+
+        return localRecoveryConfig.isLocalRecoveryEnabled()
+                        && !checkpointOptions.getCheckpointType().isSavepoint()
+                ? () ->
+                        CheckpointStreamWithResultProvider.createDuplicatingStream(
```

Comment:
I don't think this branch can (should) happen because this is the "savepoint" snapshot strategy.

Reply:
Good point!